libs/ballot-interpreter/src/hmpb-rust/timing_marks.rs

```diff
@@ -443,48 +459,130 @@ pub fn find_partial_timing_marks_from_candidate_rects(
-        .collect::<Vec<Rect>>();
-    let mut top_line = find_largest_subset_intersecting_line(
-        &top_half_rects,
-        HORIZONTAL_ANGLE,
-        MAX_BEST_FIT_LINE_ERROR,
-    );
-    let mut bottom_line = find_largest_subset_intersecting_line(
+        .collect::<Vec<_>>();
+
+    let cmp_top_line_candidates = |a: &Vec<&Rect>, b: &Vec<&Rect>| -> std::cmp::Ordering {
+        match a.len().cmp(&b.len()) {
+            std::cmp::Ordering::Equal => {
+                // if the counts are equal, sort by top opposite (where we want the smallest value)
+                average(b.iter().map(|r| r.top())).cmp(&average(a.iter().map(|r| r.top())))
+            }
+            cmp => cmp,
+        }
+    };
+
+    let cmp_bottom_line_candidates = |a: &Vec<&Rect>, b: &Vec<&Rect>| -> std::cmp::Ordering {
+        match a.len().cmp(&b.len()) {
+            std::cmp::Ordering::Equal => {
+                // if the counts are equal, sort by bottom (where we want the largest value)
+                average(a.iter().map(|r| r.bottom())).cmp(&average(b.iter().map(|r| r.bottom())))
+            }
+            cmp => cmp,
+        }
+    };
+
+    let cmp_left_line_candidates = |a: &Vec<&Rect>, b: &Vec<&Rect>| -> std::cmp::Ordering {
+        // compare by the difference from the expected number of timing marks.
+        // we can do this because the left side should be completely filled
+        let a_diff_from_expected = (a.len() as i32 - geometry.grid_size.height).abs();
+        let b_diff_from_expected = (b.len() as i32 - geometry.grid_size.height).abs();
+
+        match a_diff_from_expected.cmp(&b_diff_from_expected) {
+            std::cmp::Ordering::Equal => {
+                // if the counts are equal, sort by left opposite (where we want the smallest value)
+                average(b.iter().map(|r| r.left())).cmp(&average(a.iter().map(|r| r.left())))
+            }
+            // swap the ordering because we're using max_by and we want to minimize the diff
+            std::cmp::Ordering::Less => std::cmp::Ordering::Greater,
+            std::cmp::Ordering::Greater => std::cmp::Ordering::Less,
+        }
+    };
+
+    let cmp_right_line_candidates = |a: &Vec<&Rect>, b: &Vec<&Rect>| -> std::cmp::Ordering {
+        // compare by the difference from the expected number of timing marks.
+        // we can do this because the right side should be completely filled
+        let a_diff_from_expected = (a.len() as i32 - geometry.grid_size.height).abs();
+        let b_diff_from_expected = (b.len() as i32 - geometry.grid_size.height).abs();
+
+        match a_diff_from_expected.cmp(&b_diff_from_expected) {
+            std::cmp::Ordering::Equal => {
+                // if the counts are equal, sort by right (where we want the largest value)
+                average(a.iter().map(|r| r.right())).cmp(&average(b.iter().map(|r| r.right())))
+            }
+            // swap the ordering because we're using max_by and we want to minimize the diff
```

Comment:
The number of comments I added to this effect indicates this may be a bit of a code smell. I'm open to suggestions for improvements.

Reply:
We could, but we'd need to swap the earlier `cmp` calls.